Instrument ILM with user action telemetry

cjcenizal opened the pull request:
@bmcconaghy I wasn't able to test the "Retry lifecycle step" action, because I couldn't create an error to enable that option Index Management. Could you test this and/or give me steps on how to do that?

Actions tracked:

* "Load app" action
* "Create policy" action
* "Update policy" action
* "Delete policy" action
* "Use cold phase" action
* "Use warm phase" action
* "Use set priority" action
* "Use freeze index" action
* "Attached to index" action
* "Attached to index template" action
* "Detached from index" action
* "Re-run phase" action
* "Open edit screen" action

Telemetry shape as reported by `http://localhost:5601/api/stats?extended=true`:

```json
{
  "user_actions":[
    {
      "key":"app_load",
      "value":24
    },
    {
      "key":"policy_create",
      "value":15
    },
    {
      "key":"policy_delete",
      "value":1
    },
    {
      "key":"policy_attach_index",
      "value":2
    },
    {
      "key":"policy_attach_index_template",
      "value":1
    },
    {
      "key":"policy_detach_index",
      "value":1
    },
    {
      "key":"config_cold_phase",
      "value":2
    },
    {
      "key":"config_warm_phase",
      "value":3
    },
    {
      "key":"config_set_priority",
      "value":3
    },
    {
      "key":"config_freeze_index",
      "value":2
    },
    {
      "key":"edit_click",
      "value":3
    },
    {
      "key":"index_retry_step",
      "value":1
    }
  ]
}
```

cjcenizal committed:
Track user actions for creating, updating, and deleting and index lifecycle policy.

cjcenizal committed:
Track user actions for attaching an index, attaching an index template, and detaching an index.

cjcenizal committed:
Track app load action.

cjcenizal committed:
Refactor constants to be in root of public.

cjcenizal committed:
Add support to user actions API for comma-delimited action types.

cjcenizal committed:
Track usage of cold phase, warm phase, set priority, and freeze index.
- Rename track_user_action.js to user_action.js

cjcenizal committed:
Track retry step user action.

cjcenizal committed:
Track clicks on the policy name to edit it.

cjcenizal committed:
Remove trackUserRequest and use trackUserAction instead.

elasticmachine commented:
Pinging @elastic/es-ui

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5091/)

cjcenizal committed:
Refactor rollups trackUserRequest to be easier to understand.

cjcenizal commented on x-pack/plugins/rollup/public/crud_app/services/track_user_action.js:
```diff
@@ -13,8 +13,16 @@ export function trackUserAction(actionType) {
   // Only track successful actions.
-  request.then(() => trackUserAction(actionType));
-  return request;
+  return request.then(response => {
```
@bmcconaghy I updated the Rollups code based on some suggestions from @sebelga about how to make the code easier to follow. Do you find this clearer?

cjcenizal committed:
Add missing string type to user action API.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5163/)

bmcconaghy reviewed (dismissed):
Getting an error for retry lifecycle action:
![image](https://user-images.githubusercontent.com/2479496/54284950-57531c00-4577-11e9-9217-4275fb8c3e6b.png)
To reproduce, creation a policy with rollover (max 1 document), then assign the policy to an index without an alias and then add a document.  This will trigger rollover which will fail.  Once it does, you can retry the step.

bmcconaghy commented:
added a log statement: 
![image](https://user-images.githubusercontent.com/2479496/54285754-ef053a00-4578-11e9-9d35-7153344a3653.png)

bmcconaghy commented:
So the issue is that getHttpClient() will return null in this case.  The extension actions pass in their own httpClient, so you need to have the signature here look like this:
```
function trackUserAction(actionType, httpClient=api.getHttpClient()) {
```
and then some tracking can pass in the httpClient if they need to.

cjcenizal committed:
Use underscores instead of dashes for user action app names.

cjcenizal committed:
Switch from componentWillMount to componentDidMount/useEffect.
- Standardize use of HashRouter within app.js to avoid calling useEffect when leaving the app.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5223/)

sebelga commented on x-pack/plugins/index_lifecycle_management/common/constants/user_action.js:
```diff
@@ -0,0 +1,39 @@
+export const UA_POLICY_ATTACH_INDEX = 'policy_attach_index';
+export const UA_POLICY_ATTACH_INDEX_TEMPLATE = 'policy_attach_index_template';
+export const UA_POLICY_DETACH_INDEX = 'policy_detach_index';
+export const UA_POLICY_EDIT_CLICK = 'policy_edit_click';
```
I don't see where this action is being tracked. There is a `UA_EDIT_CLICK` being tracked but not this one.

cjcenizal replied:
You have eagle eyes!

sebelga commented on x-pack/plugins/rollup/public/crud_app/index.js:
```diff
@@ -33,9 +32,7 @@ const renderReact = async (elem) => {
-        <HashRouter>
-          <App />
-        </HashRouter>
+        <App />
```
This changes currently crashes Kibana when accessing the app because the router is needed in the `app.js` `registerRouter()`.

cjcenizal replied:
Good catch! I fixed this by using an HOC. Maybe this is something we can extract into a root-level `public` dir for global use?

loixlab reviewed (changes requested):
Great work overall @cjcenizal ! There though a change that crashes Kibana when accessing the rollup app. I left a comment. 👍

cjcenizal commented on x-pack/plugins/index_lifecycle_management/common/constants/user_action.js:
```diff
@@ -0,0 +1,39 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+export const UA_APP_NAME = 'index-lifecycle-management';
```
Done

cjcenizal commented on x-pack/plugins/index_lifecycle_management/public/app.js:
```diff
@@ -4,18 +4,27 @@
-    </Switch>
-  </HashRouter>
-);
+export class App extends Component {
```
Done

cjcenizal commented on x-pack/plugins/index_lifecycle_management/public/app.js:
```diff
@@ -4,18 +4,23 @@
-  </HashRouter>
-);
+export const App = () => {
+  useEffect(() => trackUserAction(UA_APP_LOAD, []));
```
Great catch! Thanks.

cjcenizal committed:
Fix incorrect useEffect args.

cjcenizal committed:
Fix registerRouter bug in Rollups by adding a ShareRouter HOC.

cjcenizal committed:
Remove unused UA_POLICY_EDIT_CLICK constant.

cjcenizal committed:
Pass provided httpClient through to trackUserAction.

cjcenizal commented:
@bmcconaghy @sebelga Thanks for the reviews! I've addressed your feedback, could you please review again?

Bill, I wasn't able to get a failing step with the steps you provided. To ensure the index doesn't have an alias, do I need to do anything special to it? This is what I've been doing:

```bash
PUT test
GET _aliases // returns "test": { "aliases": {} }
```

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5877/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5885/)

cjcenizal committed:
Fix Index Management jest tests by exporting AppWithoutRouter.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5897/)

loixlab reviewed (approved):
LGTM! Tested locally and working great

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6871/)

cjcenizal commented:
I'm going to merge this for now to unblock myself. @bmcconaghy When you're back could you help me reproduce a failing step so I can test the "retry" interaction?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6879/)
